dev-docs/RFCs/vNext/layer-extension-rfc.md

```diff
@@ -0,0 +1,157 @@
+
+To begin with, adding a feature to an existing core layer is no easy task. It usually involves modification to multiple steps in the layer lifecycle, e.g. adding new props, uniforms, declaring and generating new attributes, and injecting code into the shaders. It requires the author to understand the general WebGL render pipeline and GLSL, deck.gl's layer lifecycle and attribute management system, and sometimes luma.gl. This is outlined in our tutorial for [subclassed layers](/docs/developer-guide/subclassed-layers.md).
+
+Secondly, because of the ad-hoc nature of layer customization, it is difficult to package and reuse an "extension" for other layers. This is apprent in the [kepler.gl repo](https://github.com/uber/kepler.gl/tree/master/src/deckgl-layers) where a custom version of every core layer must be created in order to override a simple default bahavior.
+
+To mitigate this, we can of course add as many customizable functionalities to the base layer as possible. However, it is a major concern that inflating the base layer will sacrefice performance in return. Additional attributes takes CPU cycles to generate and memory to store. Larger shaders take longer to compile. WebGL1 has a restricting limit of 16 attributes per vertex shader, so adding attributes without the user asking for it will make the layer less extensible instead.
+
+
+## Case Studies
+
+### fp64
+
+* Shader module: `project64` (replaces `project32`)
+
+### Data Filter
+
+See [Data Filter RFC](/dev-docs/RFCs/v6.0/data-filter-rfc.md)
+
+Components:
+
+* Props: `getFilterValue`, `filterRange`
+* Shader module: `filter`
+* Attributes: `instanceFilterValue`
+* Shader injection:
+  - Vertex shader: `filter_setVisibility(instanceFilterValue);`
+  - Frament shader: `gl_FragColor = filter_filterColor(gl_FragColor);`
+
+### Brushing
+
+See [ArcBrushingLayer example](https://gist.github.com/Pessimistress/dc2becf3809c67dc443b4dbab1b9a46f#file-index-html-L128)
+
+A generic version of this functionality would need the following components:
+
+* Props: `enableBrush`, `brushRadius`
+* Shader module: `brushing`
+* Shader injection:
+  - Vertex shader: `brush_setVisibility(instancePositions);`
+  - Frament shader: `gl_FragColor = brush_filterColor(gl_FragColor);`
+* Redraw when the pointer moves (Add mouse position to the layer context? To standard uniforms?)
```

Comment:
My sense is that there should be a shader module that provides the mouse coordinates, and the extension can just depend on that module. But that is because I find module dependencies more elegant than shader parameter injections.

If that works we could just make a `mouse` shader module. But perhaps we need to add it to the project module to make sure viewports are handled correctly etc and perhaps we don't want extensions to depend on `project`...

Reply:
Touched this in another comment - it isn't an issue to pass the mouse position as a uniform, but the current architecture does not allow redrawing on mouse move. To solve this we need to probably add a new stage to the layer lifecycle, and this cannot be accomplished with a shader module.